src/Shared/Debugging/PrintLineDebugger.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using CommonWriterType = System.Action<string, string, System.Collections.Generic.IEnumerable<string>>;
+
+namespace Microsoft.Build.Shared.Debugging
+{
+    /// <summary>
+    ///     A class to help with printline debugging in difficult environments like CI, or when debugging msbuild through other
+    ///     tools like VS or CLI.
+    ///     See example usages in PrintLineDebugger_Tests
+    /// </summary>
+    internal class PrintLineDebugger : IDisposable
```

Comment:
Should we ifdef this to make sure we don't accidentally slow down released builds?

Reply:
Yes, for now it's probably for the best. Future wise, it would be nice we had something turned on all the time so we could better debug msbuild binaries already integrated in VS / CLI. A unified (as opposed to current state of having N ways) and configurable design for adding printline statements for anywhere in the code. It could be a combination of prewritten and committed groups of messages that can be turned on via unique environment variables (e.g. node communication message group, evaluation message group, scheduler message group, etc).